File: sp/reversi/server/src/main/cpp/server/connection/message/payload/payload.cpp

#include <regex>
#include "payload.hpp"
#include "../../../../utils/utils.hpp"
#include "../../consts/consts.hpp"

static const std::regex LIST_PATTERN = std::regex(R"(^\[(("[^"]*(?:","[^"]*)*")|\d+(?:,\d+)*)\]$)");
static const std::regex LIST_INT_PATTERN = std::regex(R"(^\[(\d+(?:,\d+)*)\]$)");
static const std::regex LIST_STRING_PATTERN = std::regex(R"(^\[("[^"]*(?:","[^"]*)*")\]$)");
static const std::regex STRING_PATTERN = std::regex("^\"([^\"]*)\"$");
static const std::regex INT_PATTERN = std::regex("^\\d+$");
static const std::regex BOOL_PATTERN = std::regex("^(true|false)$");
static const std::regex NULL_PATTERN = std::regex("^(null)$");

static const std::string LIST_SEPARATOR = ",";
static const std::string KEY_SEPARATOR = "=";
static const char SEPARATOR = ';';

payload::payload() {
    this->data = std::make_shared<std::map<std::string, std::shared_ptr<object>>>();
}

payload::payload(const payload* other){
    this->data = other->data;
}

void payload::set_value(const std::string &key, const std::shared_ptr<object> &value) {
    if (!validate_value(value)){
        throw std::invalid_argument("Provided value is invalid and can not be set!");
    }
    (*this->data)[key] = value;
}

std::shared_ptr<object> payload::get_value(const std::string &key) {
    auto it = this->data->find(key);
    if (it == this->data->end()){
        return nullptr;
    }

    return it->second;
}

std::shared_ptr<std::map<std::string, std::shared_ptr<object>>> payload::get_data() {
    return this->data;
}

std::string payload::to_string() {
    std::string result = "payload={";
    for (auto const& item : *this->data){
        result += item.first + "=" + item.second->to_string() + ", ";
    }

    if (result.length() > 1){
        result.replace(result.length()-2, 1, "}");
        result.replace(result.length()-1, 1, "");
    } else{
        result += "}";
    }


    return result;
}

std::shared_ptr<objects_vector> payload::parse_int_list(const std::string& value) {
    auto result = std::make_shared<objects_vector>();
    std::smatch match;
    std::regex_search(value, match, LIST_PATTERN);
    std::string tokens = match[1];
    size_t pos = 0;
    std::string token;
    while ((pos = tokens.find(LIST_SEPARATOR)) != std::string::npos) {
        token = tokens.substr(0, pos);
        result->push_back(parse_int(token));
        tokens.erase(0, pos + LIST_SEPARATOR.length());
    }
    result->push_back(parse_int(tokens)); // Last token
    return result;
}

std::shared_ptr<objects_vector> payload::parse_string_list(const std::string& value) {
    auto result = std::make_shared<objects_vector>();
    std::smatch match;
    std::regex_search(value, match, LIST_STRING_PATTERN);
    std::string tokens = match[1];
    size_t pos = 0;
    std::string token;
    while ((pos = tokens.find(LIST_SEPARATOR)) != std::string::npos) {
        token = tokens.substr(0, pos);
        result->push_back(parse_string(token));
        tokens.erase(0, pos + LIST_SEPARATOR.length());
    }
    result->push_back(parse_string(tokens)); // Last token
    return result;
}

std::shared_ptr<objects_vector> payload::parse_list(const std::string& value) {
    std::smatch str_list_match;
    std::smatch int_list_match;
    if (std::regex_search(value, str_list_match, LIST_STRING_PATTERN)) {
        return parse_string_list(value);
    } else if (std::regex_search(value, int_list_match, LIST_INT_PATTERN)) {
        return parse_int_list(value);
    } else {
        throw std::invalid_argument("Invalid list type: " + value);
    }
}

std::shared_ptr<string> payload::parse_string(const std::string& value) {
    return std::make_shared<string>(std::string(value.substr(1, value.length() - 2)));
}

std::shared_ptr<integer> payload::parse_int(const std::string& value) {
    return std::make_shared<integer>(std::stoi(value));
}

std::shared_ptr<boolean> payload::parse_boolean(const std::string& value) {
    return std::make_shared<boolean>(value == "true");
}

std::shared_ptr<object> payload::parse_value(const std::string& value) {
    std::smatch str_match;
    std::smatch int_match;
    std::smatch bool_match;
    std::smatch null_match;
    std::smatch list_match;
    if (std::regex_search(value, str_match, STRING_PATTERN)) {
        return parse_string(value);
    } else if (std::regex_search(value, int_match, INT_PATTERN)) {
        return parse_int(value);
    } else if (std::regex_search(value, bool_match, BOOL_PATTERN)) {
        return parse_boolean(value);
    } else if (std::regex_search(value, null_match, NULL_PATTERN)) {
        return nullptr;
    } else if (std::regex_search(value, list_match, LIST_PATTERN)){
        return parse_list(value);
    }else {
        throw std::invalid_argument("Invalid value: " + value);
    }
}

void payload::parse_token(const std::string& token, const std::shared_ptr<payload>& _payload) {
    size_t pos = token.find(KEY_SEPARATOR);
    if (pos != std::string::npos) {
        std::string key = token.substr(0, pos);
        std::string value = token.substr(pos + KEY_SEPARATOR.length());
        _payload->set_value(key, parse_value(value));
    } else {
        throw std::invalid_argument("Unable to parse token: " + token);
    }
}

std::shared_ptr<payload> payload::parse(const std::string& str) {
    auto _payload = std::make_shared<payload>();
    auto tokens = split_str(str, SEPARATOR);
    for (auto const& token: *tokens){
        parse_token(token, _payload);
    }
    return _payload;
}

//std::shared_ptr<payload> payload::extract(const std::string& str){
//    if (str.length() <= constants::MSG_HEADER_LENGTH){
//        return std::make_shared<payload>();
//    }
//
//    std::string payload_str = str.substr(constants::MSG_HEADER_LENGTH);
//    return parse(payload_str);
//}

std::string payload::map_string(const std::shared_ptr<string> &str){
    if (str == nullptr){
        return "null";
    }

    return '"' + str->to_string() + '"';
}

std::string payload::map_string_list(const std::shared_ptr<objects_vector>& strings_list){
    if (strings_list == nullptr || strings_list->empty()){
        return "";
    }
    std::ostringstream oss;

    for (const auto& str : *strings_list) {
        if (std::shared_ptr<string> str_ptr = std::dynamic_pointer_cast<string>(str)){
            oss << map_string(str_ptr) << ',';
        }else{
            throw std::invalid_argument("Vector contains not string type object!");
        }
    }
    std::string result = oss.str();
    result.pop_back();
    return result;
}

std::string payload::map_int(const std::shared_ptr<integer> &value) {
    if (value == nullptr){
        return "null";
    }

    return value->to_string();
}

std::string payload::map_ints_list(const std::shared_ptr<objects_vector> &integers_list) {
    if (integers_list == nullptr || integers_list->empty()){
        return "";
    }
    std::ostringstream oss;

    for (const auto& item : *integers_list) {
        if (std::shared_ptr<integer> int_ptr = std::dynamic_pointer_cast<integer>(item)){
            oss << map_int(int_ptr) << ',';
        }else{
            throw std::invalid_argument("Vector contains not integer type object!");
        }

    }
    std::string result = oss.str();
    result.pop_back();
    return result;
}

std::string payload::map_list(const std::shared_ptr<objects_vector>& list){
    if (list == nullptr || list->empty()) {
        return "null";
    }
    std::string result;
    const auto& first_element = list->at(0);
    if (std::shared_ptr<integer> int_ptr = std::dynamic_pointer_cast<integer>(first_element)) {
        result = map_ints_list(list);
    } else if (std::shared_ptr<string> str_ptr = std::dynamic_pointer_cast<string>(first_element)) {
        result = map_string_list(list);
    }else {
        throw std::invalid_argument("Unknown list items type to map.");
    }
    return "[" + result + "]";
}

std::string payload::map_boolean(const std::shared_ptr<boolean> &value) {
    return value->value() ? "true" : "false";
}

std::string payload::map_object(const std::shared_ptr<object>& obj){
    if (obj == nullptr){
        return "null";
    }
    auto obj_tmp = *obj;
    if (std::shared_ptr<string> str_ptr = std::dynamic_pointer_cast<string>(obj)){
        return map_string(str_ptr);
    }
    if (std::shared_ptr<integer> int_ptr = std::dynamic_pointer_cast<integer>(obj)){
        return map_int(int_ptr);
    }
    if (std::shared_ptr<objects_vector> vector_ptr = std::dynamic_pointer_cast<objects_vector>(obj)){
        return map_list(vector_ptr);
    }
    if (std::shared_ptr<boolean> bool_ptr = std::dynamic_pointer_cast<boolean>(obj)){
        return map_boolean(bool_ptr);
    }
    return "";
}

std::string payload::map(const std::shared_ptr<payload> &_payload) {
    if (_payload->get_data()->empty()){
        return "";
    }
    std::ostringstream oss;
    for (auto const& it : *(_payload->get_data())){
        oss << it.first << '=' << map_object(it.second) << SEPARATOR;
    }
    oss.seekp(-1, std::ios_base::end); // Move back to remove the last separator
    return oss.str();
}

std::string payload::construct() const{
    //TODO: replace with function overloading
    return map(std::make_shared<payload>(this));
}

bool payload::validate_str_value(const std::shared_ptr<string> value) {
    if (value == nullptr){
        return true;
    }
    return value->find(SEPARATOR) == std::string::npos;
}

bool payload::validate_str_list_value(const std::shared_ptr<objects_vector>& value){
    if (value == nullptr || value->empty()){
        return true;
    }
    for (auto str : *value){
        if (auto str_ptr = std::dynamic_pointer_cast<string>(str)){
            if (!validate_str_value(str_ptr)){
                return false;
            }
        } else {
            throw std::invalid_argument("Vector contains not string type object!");
        }
    }

    return true;
}

bool payload::validate_list_value(const std::shared_ptr<objects_vector>& value) {
    if (value == nullptr || value->empty()){
        return true;
    }
    auto first_item = value->at(0);
    if (auto str_ptr = std::dynamic_pointer_cast<string>(first_item)){
        return validate_str_list_value(value);
    }

    //other types are valid
    return true;
}

bool payload::validate_value(const std::shared_ptr<object>& value){
    if (value == nullptr){
        return true;
    }

    if (auto str_ptr = std::dynamic_pointer_cast<string>(value)){
        return validate_str_value(str_ptr);
    }
    if (auto vector_ptr = std::dynamic_pointer_cast<objects_vector>(value)){
        return validate_list_value(vector_ptr);
    }

    //other types are valid
    return true;
}

File: sp/reversi/client/composeApp/src/commonMain/kotlin/com/vpavlov/ups/reversi/client/connection/message/Payload.kt
package com.vpavlov.ups.reversi.client.connection.message

import com.vpavlov.ups.reversi.client.connection.MSG_HEADER_LENGTH
import java.util.regex.Matcher
import java.util.regex.Pattern


private val ALLOWED_TYPES = setOf(
    String::class,
    Int::class,
    Boolean::class,
    List::class
)

private const val LIST_START = '['

private const val LIST_END = ']'

private const val SEPARATOR = ';'

private var KEY_SEPARATOR = "="

private var LIST_SEPARATOR = ","


val LIST_PATTERN: Pattern = Pattern.compile("^\\[((\"[^\"]*(?:\",\"[^\"]*)*\")|\\d+(?:,\\d+)*)]$")

val LIST_INT_PATTERN: Pattern = Pattern.compile("^\\[(\\d+(?:,\\d+)*)]$")

val LIST_STRING_PATTERN: Pattern = Pattern.compile("^\\[(\"[^\"]*(?:\",\"[^\"]*)*\")]$")

val STRING_PATTERN: Pattern = Pattern.compile("^\"([^\"]*)\"$")

val INT_PATTERN: Pattern = Pattern.compile("^\\d+$")

val BOOL_PATTERN: Pattern = Pattern.compile("^(true|false)$")

val NULL_PATTERN: Pattern = Pattern.compile("^(null)$")



data class Payload(val data: MutableMap<String, Any?> = mutableMapOf()) {

    @Synchronized
    fun setValue(key: String, value: Any?) {
        require(key.isNotBlank()) {
            "key could not be blank."
        }
        if (value == null){
            data[key] = null
            return;
        }
        require(value::class in ALLOWED_TYPES) {
            "value should be of type: $ALLOWED_TYPES."
        }
        require(value is List<*> && value.all { it in ALLOWED_TYPES }) {
            "value should be of type: $ALLOWED_TYPES."
        }
        data[key] = value
    }

    @Synchronized
    fun getValue(key: String): Any? = data[key]

    @Synchronized
    fun construct(): String {
        if (data.isEmpty()) {
            return ""
        }
        val sb = StringBuilder()
        data.forEach { (key, value) ->
            sb.append(key).append(KEY_SEPARATOR)
            sb.append(map(value)).append(SEPARATOR)
        }
        return sb.toString()
    }

    private fun map(value: Any?): String?{
        if (value == null){
            return "null"
        }
        return when(value){
            is String -> mapString(value)
            is Int -> mapInt(value)
            is Boolean -> mapBoolean(value)
            is List<*> -> mapList(value)
            else -> null
        }
    }

    private fun mapString(value: String): String = "\"$value\""

    private fun mapStringsList(value: List<String>): String{
        val sb = java.lang.StringBuilder()
        for (string in value) {
            sb.append(mapString(string)).append(',')
        }
        sb.deleteCharAt(sb.length - 1)
        return sb.toString()
    }

    private fun mapInt (value: Int) = value.toString()

    private fun mapIntList(value: List<Int>): String {
        val sb = java.lang.StringBuilder()
        for (integer in value) {
            sb.append(mapInt(integer)).append(',')
        }

        //delete last comma
        sb.deleteCharAt(sb.length - 1)
        return sb.toString()
    }

    private fun mapList(value: List<*>): String {
        if (value.isEmpty()){
            return "$LIST_START$LIST_END"
        }
        val result = when(value[0]){
            is Int -> mapIntList(value as List<Int>)
            is String -> mapStringsList(value as List<String>)
            else -> ""
        }
        return "$LIST_START$result$LIST_END"
    }

    private fun mapBoolean(value: Boolean) = value.toString()

    companion object{
        fun parse(message: String): Payload{
            if (message.length <= MSG_HEADER_LENGTH){
                return Payload()
            }
            val payloadStr = message.substring(MSG_HEADER_LENGTH)
            return parseHelper(payloadStr)
        }

        private fun parseHelper(payloadStr: String): Payload {
            val payload = Payload()
            if (payloadStr.isBlank()){
                return payload
            }
            val payloadTokens = payloadStr.split(SEPARATOR)
            payloadTokens.forEach { token->
                parseToken(token, payload)
            }
            return payload
        }

        private fun parseToken(token: String, payload: Payload){
            val keyValueTokens = token.split(KEY_SEPARATOR)
            require (keyValueTokens.size == 2){
                "Unable to parse token: $token"
            }
            val key = keyValueTokens[0]
            require(key.isNotBlank()){
                "key could not be blank."
            }
            val value = keyValueTokens[1]
            payload.setValue(key,parseValue(value))
        }

        private fun parseValue(value: String): Any? {
            val trimmed = value.trim()
            val nullMatcher: Matcher = NULL_PATTERN.matcher(trimmed)
            if (nullMatcher.matches()) {
                return null
            }
            val listMatcher: Matcher = LIST_PATTERN.matcher(trimmed)
            if (listMatcher.matches()) {
                return parsePayloadList(trimmed)
            }
            return parsePayloadValue(trimmed)
        }

        private fun parsePayloadValue(value: String): Any {
            if (STRING_PATTERN.matcher(value).matches()) {
                return parsePayloadString(value)
            }
            if (INT_PATTERN.matcher(value).matches()) {
                return parsePayloadInt(value)
            }
            if ( BOOL_PATTERN.matcher(value).matches()) {
                return value.toBoolean()
            }
            throw IllegalArgumentException("Invalid value: $value")
        }

        private fun parsePayloadString(value: String) = value.substring(1,value.length-1)

        private fun parsePayloadInt(value: String) = Integer.parseInt(value)

        private fun parsePayloadList(value: String): List<Any> {
            if (LIST_STRING_PATTERN.matcher(value).matches()) {
                return parseStringList(value)
            }
            if (LIST_INT_PATTERN.matcher(value).matches()) {
                return parseIntList(value)
            }
            throw java.lang.IllegalArgumentException("Invalid list type: $value")
        }

        private fun parseStringList(value: String): List<String> {
            val trimmed = value.substring(1, value.length - 1)
            val result = mutableListOf<String>()
            val tokens = trimmed.split(LIST_SEPARATOR)
            for (token in tokens) {
                result.add(parsePayloadString(token))
            }
            return result
        }

        private fun parseIntList(value: String): List<Int> {
            val trimmed = value.substring(1, value.length - 1)
            val result = mutableListOf<Int>()
            val tokens = trimmed.split(LIST_SEPARATOR)
            for (token in tokens) {
                result.add(parsePayloadInt(token))
            }
            return result
        }
    }

}